// cycle-discord-google-api/dist/index.js
"use strict";

function GoogleApi(sources) {
    var result$ = sources.HTTP.combineLatest(sources.props, function (res$, _ref) {
        var URL = _ref.URL;
        return { res$: res$, URL: URL };
    }).filter(function (_ref2) {
        var res$ = _ref2.res$;
        var URL = _ref2.URL;
        return res$.request.url.indexOf(URL) !== -1;
    }).map(function (_ref3) {
        var res$ = _ref3.res$;
        return res$;
    }).mergeAll().map(function (res) {
        return res.body;
    }).map(function (_ref4) {
        var _ref4$items = _ref4.items;
        var items = _ref4$items === undefined ? [] : _ref4$items;
        return items[Math.floor(Math.random() * items.length)];
    });
    var triggerMessage$ = sources.Discord.on(sources.props.map(function (_ref5) {
        var trigger = _ref5.trigger;
        return trigger;
    }));
    return {
        Discord: result$.combineLatest(sources.props, function (result, _ref6) {
            var transform = _ref6.transform;

            if (!result) {
                return "No result found";
            }
            return transform(result);
        }).zip(triggerMessage$, function (result, message) {
            return {
                content: result,
                channelId: message.channel.id
            };
        }),
        HTTP: triggerMessage$.combineLatest(sources.props, function (message, _ref7) {
            var URL = _ref7.URL;
            var trigger = _ref7.trigger;
            return URL + "&q=" + encodeURIComponent(trigger.exec(message.content)[1]);
        })
    };
}
exports.GoogleApi = GoogleApi;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = GoogleApi;
exports.IMAGE_SEARCH_URL = "https://www.googleapis.com/customsearch/v1?num=10&safe=off&searchType=image";
exports.YOUTUBE_SEARCH_URL = "https://www.googleapis.com/youtube/v3/search?part=snippet&maxResults=1&order=relevance&safeSearch=none";
