// cycle-discord-google-api/src/index.ts
import {Sources, Sink} from 'cycle-discord-google-api'

export function GoogleApi(sources: Sources): Sink {
    const result$ = sources.HTTP
        .combineLatest(sources.props, (res$, {URL}) => ({res$, URL}))
        .filter(({res$, URL}) => res$.request.url.indexOf(URL) !== -1)
        .map(({res$}) => res$)
        .mergeAll()
        .map(res => res.body)
        .map(({items = []}) => items[Math.floor(Math.random() * items.length)])

    const triggerMessage$ = sources.Discord.on(sources.props.map(({trigger}) => trigger))

    return {
        Discord: result$
            .combineLatest(sources.props, (result, {transform}) => {
                if (!result) {
                    return `No result found`
                }

                return transform(result)
            })
            .zip(triggerMessage$, (result, message) => ({
                content: result,
                channelId: message.channel.id
            })),
        HTTP: triggerMessage$
            .combineLatest(sources.props, (message, {URL, trigger}) =>
                `${URL}&q=${encodeURIComponent(trigger.exec(message.content)[1])}`
            )
    }
}

export default GoogleApi

export const IMAGE_SEARCH_URL = `https://www.googleapis.com/customsearch/v1?num=10&safe=off&searchType=image`
export const YOUTUBE_SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&maxResults=1&order=relevance&safeSearch=none`
